tower-balance/src/p2c/service.rs

```diff
@@ -104,82 +77,39 @@ where
             match try_ready!(self.discover.poll().map_err(|e| error::Discover(e.into()))) {
                 Change::Remove(key) => {
                     trace!("remove");
-                    self.evict(&key)
+                    self.services.evict(&key);
                 }
                 Change::Insert(key, svc) => {
                     trace!("insert");
-                    self.evict(&key);
-                    self.push_unready(key, svc);
+                    // If this service already existed in the set, it will be
+                    // replaced as the new one becomes ready.
+                    self.services.push_service(key, svc);
                 }
             }
         }
     }
 
-    fn push_unready(&mut self, key: D::Key, svc: D::Service) {
-        let (tx, rx) = oneshot::channel();
-        self.cancelations.insert(key.clone(), tx);
-        self.unready_services.push(UnreadyService {
-            key: Some(key),
-            ready: Ready::new(svc),
-            cancel: rx,
-        });
-    }
-
-    fn evict(&mut self, key: &D::Key) {
-        // Update the ready index to account for reordering of ready.
-        if let Some((idx, _, _)) = self.ready_services.swap_remove_full(key) {
-            self.next_ready_index = self
-                .next_ready_index
-                .and_then(|i| Self::repair_index(i, idx, self.ready_services.len()));
-            debug_assert!(!self.cancelations.contains_key(key));
-        } else if let Some(cancel) = self.cancelations.remove(key) {
-            let _ = cancel.send(());
-        }
-    }
-
     fn poll_unready(&mut self) {
+        // Note: this cannot perturb the order of ready services.
```

Comment:
I'm not sure I follow this comment.

Reply:
This isn't relevant in the balancer anymore. This was referring to the fact that poll_pending will only interact with the pending set and not the ready set, so the next-ready-index can't be impacted by this...